internal/tiltfile/tiltfile_state.go

```diff
@@ -289,6 +297,80 @@ func (s *tiltfileState) assembleK8s() error {
 
 }
 
+func (s *tiltfileState) assembleK8sV2() error {
+	err := s.assembleK8sByWorkload()
+	if err != nil {
+		return err
+	}
+
+	err = s.assembleK8sUnresourced()
+	if err != nil {
+		return err
+	}
+
+	for _, r := range s.k8s {
+		if err := s.validateK8s(r); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *tiltfileState) assembleK8sByWorkload() error {
+	var workloads, rest []k8s.K8sEntity
+	for _, e := range s.k8sUnresourced {
+		isWorkload, err := s.isWorkload(e)
+		if err != nil {
+			return err
+		}
+		if isWorkload {
+			workloads = append(workloads, e)
+		} else {
+			rest = append(rest, e)
+		}
+	}
+	s.k8sUnresourced = rest
+
+	resourceNames, err := uniqueResourceNames(workloads)
+	if err != nil {
+		return err
+	}
+	for i, resourceName := range resourceNames {
+		workload := workloads[i]
+		res, err := s.makeK8sResource(resourceName)
+		if err != nil {
+			return err
+		}
+		err = res.addEntities([]k8s.K8sEntity{workload}, s.imageJSONPaths)
+		if err != nil {
+			return err
+		}
+
+		match, rest, err := k8s.FilterByMatchesPodTemplateSpec(workload, s.k8sUnresourced)
```

Comment:
can you add comments on what this code does? it took me a while to figure it out